### PR TITLE
NOISSUE - Implementation of  Group Delete

### PR DESCRIPTION
**Pull request by arvindh123**
# What type of PR is this?
Implements Group Delete

## What does this do?
Remove group and its related policies 

## Which issue(s) does this PR fix/relate to?
Example: 
Delete group: 
```bash 
curl --location --request DELETE 'http://localhost/groups/02772152-1e87-4d92-9752-19ab3c778a92' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <TOKEN>'
```

Delete channel: 
```bash
curl --location --request DELETE 'http://localhost/channels/fe5ad766-f316-43bc-8c5e-baddb4fb3b6f' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'Authorization: Bearer <TOKEN>' 
````

- Related Issue #
- Resolves #

## Have you included tests for your changes?
No, Not planned in this PR

## Did you document any new/modified feature?
No, Not Planned in this PR

### CODE REVIEW

File: internal/groups/events/streams.go [hunk 1]
```diff
@@ -172,3 +172,15 @@ func (es eventStore) delete(ctx context.Context, group groups.Group) (groups.Gro
 
 	return group, nil
 }
+
+func (es eventStore) DeleteGroup(ctx context.Context, token, id string) error {
+	if err := es.svc.DeleteGroup(ctx, token, id); err != nil {
+		return err
+	}
+
+	_, err := es.delete(ctx, groups.Group{ID: id})
```
Comment (rodneyosodo) [inline internal/groups/events/streams.go:181]: This should have a different operation name

File: auth/api/grpc/requests.go [hunk 2]
```diff
@@ -94,17 +94,17 @@ type policyReq struct {
 }
 
 func (req policyReq) validate() error {
-	if req.Subject == "" || req.SubjectType == "" {
-		return apiutil.ErrMissingPolicySub
-	}
+	// if req.Subject == "" || req.SubjectType == "" {
```
Comment (rodneyosodo) [inline auth/api/grpc/requests.go:97]: Uncomment

File: internal/groups/postgres/groups.go [hunk 3]
```diff
@@ -303,6 +303,24 @@ func (repo groupRepository) UnassignParentGroup(ctx context.Context, parentGroup
 	return nil
 }
 
+func (repo groupRepository) UnassignParentInChildren(ctx context.Context, parentGroupID string) error {
+	q := "UPDATE groups AS g SET parent_id = NULL WHERE g.parent_id = $1 ;"
```
Comment (rodneyosodo) [inline internal/groups/postgres/groups.go:307]: Shouldn't this be done at the database level with something like a trigger?

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: internal/groups/postgres/groups.go (hunk 3 above)
Comment (arvindh123) [reply internal/groups/postgres/groups.go:307]: `FOREIGN KEY (parent_id) REFERENCES groups (id) ON DELETE SET NULL`

`ON DELETE SET NULL` will be replacement for this function

Comment (arvindh123) [reply internal/groups/postgres/groups.go]: Resolved

File: auth/api/grpc/requests.go (hunk 2 above)
Comment (arvindh123) [reply auth/api/grpc/requests.go:97]: Created issue, I will do in seperate PR https://github.com/absmach/magistrala/issues/170

Comment (rodneyosodo) [reply auth/api/grpc/requests.go:97]: You can remove the code block and let it return nil

File: internal/groups/events/events.go [hunk 4]
```diff
@@ -20,6 +20,7 @@ const (
 	groupViewPerms       = groupPrefix + "view_perms"
 	groupList            = groupPrefix + "list"
 	groupListMemberships = groupPrefix + "list_by_user"
+	groupDelete          = groupPrefix + "delete"
```
Comment (rodneyosodo) [inline internal/groups/events/events.go:23]: Also change `groupRemove` to `groupChangeStatus`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: internal/groups/events/events.go (hunk 4 above)
Comment (arvindh123) [reply internal/groups/events/events.go:23]: done

File: auth/api/grpc/requests.go (hunk 2 above)
Comment (arvindh123) [reply auth/api/grpc/requests.go]: done

File: bootstrap/events/consumer/streams.go [hunk 5]
```diff
@@ -49,8 +49,8 @@ func (es *eventHandler) Handle(ctx context.Context, event events.Event) error {
 	case channelUpdate:
 		uce := decodeUpdateChannel(msg)
 		err = es.handleUpdateChannel(ctx, uce)
-	case channelRemove:
-		rce := decodeRemoveChannel(msg)
+	case channelDelete:
+		rce := decodeDeleteChannel(msg)
```
Comment (rodneyosodo) [inline bootstrap/events/consumer/streams.go:53]:
```suggestion
		dce := decodeDeleteChannel(msg)
```

File: internal/groups/events/events.go [hunk 6]
```diff
@@ -15,17 +15,18 @@ const (
...
 )
 
 var (
 	_ events.Event = (*createGroupEvent)(nil)
 	_ events.Event = (*updateGroupEvent)(nil)
-	_ events.Event = (*removeGroupEvent)(nil)
```
Comment (rodneyosodo) [inline internal/groups/events/events.go:28]: Also add a type check for  `deleteGroupEvent`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Also, the internal implementation of groups should take `streamID` as a parameter so that when it is being used for users case it is `magistrala.users` and for things it is `magistrala.things`.

Also SDK and CLI functions are missing

File: cli/channels.go [hunk 7]
```diff
@@ -81,6 +81,25 @@ var cmdChannels = []cobra.Command{
...
+		Long: "Delete channel by id.\n" +
+			"Usage:\n" +
+			"\tmagistrala-cli channels delete <channel_id> $USERTOKEN - delete the given channel ID\n",
+		Run: func(cmd *cobra.Command, args []string) {
+			if len(args) != 2 {
+				logUsage(cmd.Use)
+				return
+			}
+			err := sdk.DeleteChannel(args[0], args[1])
+			if err != nil {
```
Comment (rodneyosodo) [inline cli/channels.go:96]: Inline

File: cli/groups.go [hunk 8]
```diff
@@ -141,6 +141,25 @@ var cmdGroups = []cobra.Command{
...
+		Short: "Delete group",
+		Long: "Delete group by id.\n" +
+			"Usage:\n" +
+			"\tmagistrala-cli groups delete <group_id> $USERTOKEN - delete the given group ID\n",
+		Run: func(cmd *cobra.Command, args []string) {
+			if len(args) != 2 {
+				logUsage(cmd.Use)
+				return
+			}
+			err := sdk.DeleteGroup(args[0], args[1])
```
Comment (rodneyosodo) [inline cli/groups.go:155]: Same

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

**Review by rodneyosodo — APPROVED** [review]
LGTM

File: internal/groups/service.go [hunk 9]
```diff
@@ -601,6 +601,59 @@ func (svc service) Unassign(ctx context.Context, token, groupID, relation, membe
...
+	}
+
+	// Remove policy from domain
+	if _, err := svc.auth.DeletePolicy(ctx, &magistrala.DeletePolicyReq{
+		SubjectType: auth.DomainType,
+		Object:      groupID,
+		ObjectType:  auth.GroupType,
+	}); err != nil {
+		return err
+	}
```
Comment (dborovcanin) [inline internal/groups/service.go:638]: Is it possible to execute these in a single request to the Auth? I.e. create a separate gRPC endpoint for this?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: internal/groups/service.go (hunk 9 above)
Comment (dborovcanin) [reply internal/groups/service.go:638]: Opened relevant issue: #188.

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**